casper/src/main/scala/coop/rchain/casper/BlockCreator.scala

```diff
@@ -107,20 +107,33 @@ object BlockCreator {
       currentBlockNumber  = maxBlockNumber + 1
       earliestBlockNumber = currentBlockNumber - expirationThreshold
       deploys             = state.deployHistory
-      validDeploys = deploys.filter(
+      distinctDeploys = deploys
+        .groupBy(deploy => (deploy.deployer, deploy.timestamp))
+        .map(_._2.head)
+      validDeploys = distinctDeploys.filter(
         d => notFutureDeploy(currentBlockNumber, d) && notExpiredDeploy(earliestBlockNumber, d)
       )
-      deploysInCurrentChain <- DagOperations
-                                .bfTraverseF[F, BlockMessage](parents.toList)(
-                                  b =>
-                                    ProtoUtil
-                                      .unsafeGetParentsAboveBlockNumber[F](b, earliestBlockNumber)
-                                )
-                                .map { b =>
-                                  ProtoUtil.deploys(b).flatMap(_.deploy)
-                                }
-                                .toList
-    } yield (validDeploys -- deploysInCurrentChain.flatten).toSeq
+      deploysInCurrentChainList <- DagOperations
+                                    .bfTraverseF[F, BlockMessage](parents.toList)(
+                                      b =>
+                                        ProtoUtil
+                                          .unsafeGetParentsAboveBlockNumber[F](
+                                            b,
+                                            earliestBlockNumber
+                                          )
+                                    )
+                                    .map { b =>
+                                      ProtoUtil.deploys(b).flatMap(_.deploy)
+                                    }
+                                    .toList
+      deploysInCurrentChain = deploysInCurrentChainList.flatten.toSet
```

Comment:
if you remove the toList and the flatten you get the same result

Reply:
Isn't the flatten necessary? Does toSet also flatten?